keps/sig-api-machinery/4006-transition-spdy-to-websockets/README.md

```diff
@@ -0,0 +1,1199 @@
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+Some Kubernetes clients need to communicate with the API Server using a bi-directional
+streaming protocol, instead of the standard HTTP request/response mechanism. A streaming
+protocol provides the ability to read and write arbitrary data messages between the
+client and server, instead of providing a single response to a client request.
+For example, the commands `kubectl exec`, `kubectl attach`, and `kubectl port-forward`
+all benefit from a bi-directional streaming protocol (`kubectl cp` is build on top
+of `kubectl exec` primitives so it utilizes streaming as well). Currently,
+the bi-directional streaming solution for these `kubectl` commands is SPDY/3.1. For
+the communication leg between `kubectl` and the API Server, this enhancement transitions
+the bi-directional streaming protocol to WebSockets from SPDY/3.1.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The SPDY streaming protocol has been deprecated since 2015, and by now
+many proxies, gateways, and load-balancers do not support SPDY. Our effort to modernize
+the streaming protocol between Kubernetes clients and the API Server using WebSockets
+is necessary to enable the aforementioned intermediaries. WebSockets is a currently
+supported standardized protocol (https://www.rfc-editor.org/rfc/rfc6455) that guarantees
+compatibility and interoperability with the different components and programming
+languages. Finally, WebSockets is preferrable to HTTP/2.0 because the updated HTTP
+standard does not support streaming well. The decision to forego HTTP/2.0 is discussed
+at greater length in the [Alternatives Section](##Alternatives).
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+1. Transition the bi-directional streaming protocol from SPDY/3.1 to WebSockets for
+`kubectl exec`, `kubectl attach`, and `kubectl cp` for the communication leg
+between `kubectl` and the API Server.
+
+2. Transition the bi-directional streaming protocol from SPDY/3.1 to WebSockets
+for `kubectl port-forward` for the communication leg between `kubectl` and the API
+Server (alpha in v1.29).
+
+3. Extend the WebSockets communication leg from the API Server to Kubelet
```

Comment:
Given detection in the kube-apiserver that is similar to kubectl, the kube-apiserver to kubelet connection could switch using the same mechanism.  The auto-detection is important since kubelets can skew heavily, but that detection needs to happen regardless.

Is there a reason similar detection and patterns cannot work for kube-apiserver to kubelet cannot work?

Reply:
I have modified the KEP to schedule extending WebSocket to the Kubelet from `Post-GA` to `Pre-GA`. Please let me know if this addresses your concern.

You are correct that the mechanism for delegating to the `StreamTranslatorProxy` (which implements the WebSockets) is the same in the API Server as it is in the Kubelet, since both the API Server and the Kubelet are using the same `UpgradeAwareProxy` currently. Extending WebSockets to Kubelet will be 1) adding the same conditional delegation to `StreamTranslatorProxy` into the Kubelet, and 2) adding a step in the API Server to see if the WebSockets upgrade succeeded (101 Switching Protocols) signaling Kubelet supports the `StreamTranslatorProxy`. I think this step should be pretty straightforward.